cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,99 @@
+
+Any DNS-based service discovery solution for Kubernetes must support the
+queries described below to be considered compliant with this specification.
+
+In the query names below, values not in angle brackets, `< >`,
+are literals. The meaning of the values in angle brackets are defined
+below or in the description of the specific query response.
+
+`<zone>` = configured cluster domain, e.g. cluster.local
+
+`<ns>` = a Namespace
+
+All comparisons between query names and data in Kubernetes is **case-insensitive**.
+
+In the results column, the following definitions should be used for
+words in _italics_.
+
+- _hostname_
+
+  - For endpoints, in order of precedence:
+    - The value of the endpoint’s `hostname` field.
+    - The value of the `HostName` field stored in the map indexed by the endpoint’s IP address in the `endpoints.beta.kubernetes.io/hostnames-map` annotation.
+    - A unique, system-assigned identifier for the endpoint. The exact format and source of this identifier is not prescribed by this specification. However, it must be possible to use this to identify a specific endpoint in the context of a Service. This is used for SRV queries below in the event no explicit endpoint hostname is defined by the above methods.
+
+- _ready_
+
+ - An endpoint is considered _ready_ if its address is in the `addresses` field of the EndpointSubset object, or the corresponding service has the `service.alpha.kubernetes.io/tolerate-unready-endpoints` annotation set to `true`.
+
+The supported queries and expected responses follow.
+
+- **Type:** A, **Name:** `<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, then the result will be a single A record with the specified name and the ClusterIP of that Service.
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP (i.e., a headless service)  and there are no _ready_ endpoints for that Service, then the result will be empty (i.e., an empty Answer section).
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP and there are _ready_ endpoints for the Service, the result will be an A record with the specified name and IP for each _ready_ endpoint.
+
+- **Type:** A, **Name:** `<hostname>.<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` and a _ready_ endpoint exists for that Service with _hostname_ of `<hostname>` the result will be a single A record with the IP address of that endpoint. If no such endpoint exists, the result will be `NXDOMAIN`.
+
+- **Type:** A, **Name:** `<a>-<b>-<c>-<d>.<ns>.pod.<zone>`
+
+  - If a Pod exists with IP address `<a>.<b>.<c>.<d>`, the result will be an A record with the specified name and the IP address `<a>.<b>.<c>.<d>`.
+
+  - If no such Pod exists, the result will be `NXDOMAIN`.
```

Comment:
Current kube-dns behavior is to ignore the state of the cluster and simply reply to every query of this form with <a>.<b>.<c>.<d>, whether anything is at that IP or not.